### PR TITLE
Add validate method to ECX keymanager

### LINKED ISSUE #11619: EVP_PKEY_private_check returns 1 for Ed25519 public key
master branch

When importing just the public key for a Ed25519 key, EVP_PKEY_private_check function returns 1, it is expected to return 0.

Pseudo code:
```
unsigned char publicKey[32];
OSSL_PARAM fromdata_params[] = {
  OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_PUB_KEY, publicKey, ED25519_KEYSIZE),
	OSSL_PARAM_END
};
	
EVP_PKEY_CTX ctx1 = EVP_PKEY_CTX_new_from_name(NULL, SN_ED25519, NULL);
	
EVP_PKEY_key_fromdata_init(ctx1);
EVP_PKEY* pkey = NULL;
EVP_PKEY_fromdata(ctx1, &pkey, param);
	
EVP_PKEY_CTX ctx = EVP_PKEY_CTX_new_from_pkey(NULL, pkey, "");
assert(EVP_PKEY_private_check(ctx) != 1);
```

The problem seems to be that there is no validate implementation and therefore the function always returns 1:
```
int evp_keymgmt_validate(const EVP_KEYMGMT *keymgmt, void *keydata,
                         int selection)
{
    /* We assume valid if the implementation doesn't have a function */
    if (keymgmt->validate == NULL)
        return 1;
    return keymgmt->validate(keydata, selection);
}
```

### PR DESCRIPTION
**Pull request by slontis**
Fixes #11619

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

This needs a test to be added..

### CODE REVIEW

File: providers/implementations/keymgmt/ecx_kmgmt.c [hunk 1]
```diff
@@ -670,6 +674,104 @@ void *ecx_load(const void *reference, size_t reference_sz)
...
+static int ecx_key_public_check(const ECX_KEY *ecx, size_t len)
+{
+    return ecx->haspubkey && len == ecx->keylen;
+}
+
+static int ecx_key_private_check(const ECX_KEY *ecx, int type, size_t len)
+{
+    const unsigned char *priv = ecx->privkey;
+
+    if (priv == NULL || len != ecx->keylen)
```
Comment (mattcaswell) [inline providers/implementations/keymgmt/ecx_kmgmt.c:686]: We could even consider an `!ossl_assert` on the `len != ecx->keylen` case. This should never happen.

**Review by mattcaswell — COMMENTED** [review]
Looks good so far.

File: providers/implementations/keymgmt/ecx_kmgmt.c (hunk 1 above)
Comment (slontis) [reply providers/implementations/keymgmt/ecx_kmgmt.c]: Done..

**Comment by slontis** [comment]
@mattcaswell Tthe extra tests I had for certain bits being set in the first and last bytes of the private key failed since the RFC test vectors are just random numbers.. it does make me wonder if ecx_gen() should really be changing the bits for the X25519 and X448 types.. (Maybe this should only be done in the op that uses them??)

**Comment by kaduk** [comment]
>  it does make me wonder if ecx_gen() should really be changing the bits for the X25519 and X448 types.. (Maybe this should only be done in the op that uses them??)

As counterintuitive as it seems to me, your suggestion to only do the bit-twiddling in the op that uses them seems consistent with RFC 7748:

   Scalars are assumed to be randomly generated bytes.  For X25519, in
   order to decode 32 random bytes as an integer scalar, set the three
   least significant bits of the first byte and the most significant bit
   of the last to zero, set the second most significant bit of the last
   byte to 1 and, finally, decode as little-endian.  This means that the
   resulting integer is of the form 2^254 plus eight times a value
   between 0 and 2^251 - 1 (inclusive).  Likewise, for X448, set the two
   least significant bits of the first byte to 0, and the most
   significant bit of the last byte to 1.  This means that the resulting
   integer is of the form 2^447 plus four times a value between 0 and
   2^445 - 1 (inclusive).

This is in the definition of the X25519() and X448() functions, and thus implies that the input to those functions (including private keys) is supposed to be unmodified random bytes.

**Comment by romen** [comment]
Yes, for both X25519 and Ed25529 (and the 448) the "private key" (that somewhere else is called "seed" in the context of EdDSA) is just a string of bytes with no requirement other than the length.

The "pruning" of the bottom and top bits only happens during the operations that use this raw private key.

**Comment by slontis** [comment]
ping

**Comment by slontis** [comment]
ping

File: test/evp_pkey_provided_test.c [hunk 2]
```diff
@@ -272,6 +272,20 @@ static int test_print_key_using_encoder(const char *alg, const EVP_PKEY *pk)
...
 }
 
+#ifndef OPENSSL_NO_EC
+static int test_print_key_using_encoder_public(const char *alg,
+                                               const EVP_PKEY *pk)
+{
+    int i;
+    int ret = 1;
+
+    for (i = PUB_TEXT; i < 6; i++)
```
Comment (mattcaswell) [inline test/evp_pkey_provided_test.c:282]: Would it be better to say `<= PUB_DER` here rather than `6`?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: test/evp_pkey_provided_test.c (hunk 2 above)
Comment (slontis) [reply test/evp_pkey_provided_test.c:282]: Updated.. Updated the function above that also does this as well as put a comment above the defines saying not to modify them.

**Comment by slontis** [comment]
ping

**Review by mattcaswell — APPROVED** [review]

**Comment by slontis** [comment]
Thanks. Merged to master.

**Pull request closed without merging**